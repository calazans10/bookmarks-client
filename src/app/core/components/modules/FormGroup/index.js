import React, { useState } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import FormLabel from './atoms/FormLabel';
import FormInput from './atoms/FormInput';
import FormError from './atoms/FormError';
import './index.scss';

const FormGroup = ({ input, meta, label, type, mask, formatChars, autoFocus, disabled }) => {
  const [isFocused, setIsFocused] = useState(false);

  const isInvalid = meta.touched && meta.error && !isFocused;

  const divClass = classNames({
    'form-group': true,
    'form-group--invalid': isInvalid,
  });

  const onFocus = () => {
    setIsFocused(true);
  };

  const onBlur = () => {
    setIsFocused(false);
  };

  return (
    <div className={divClass}>
      <FormLabel htmlFor={input.name}>{label}</FormLabel>
      <FormInput
        input={input}
        type={type}
        mask={mask}
        formatChars={formatChars}
        autoFocus={autoFocus}
        disabled={disabled}
        placeholder={label}
        onFocus={onFocus}
        onBlur={onBlur}
      />
      {isInvalid && <FormError>{meta.error}</FormError>}
    </div>
  );
};

FormGroup.propTypes = {
  input: PropTypes.object.isRequired,
  meta: PropTypes.object.isRequired,
  label: PropTypes.string,
  type: PropTypes.string,
  mask: PropTypes.string,
  formatChars: PropTypes.object,
  autoFocus: PropTypes.bool,
  disabled: PropTypes.bool,
};

FormGroup.defaultProps = {
  label: '',
  type: 'tel',
  mask: '',
  formatChars: { 9: '[0-9]', '?': '[0-9]' },
  autoFocus: false,
  disabled: false,
};

export default FormGroup;
