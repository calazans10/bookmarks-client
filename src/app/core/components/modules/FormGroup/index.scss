@import 'styles/utilities';

.form-group {
  margin-bottom: rem(20);

  &__label {
    display: block;
    margin-bottom: rem(5);
    font-weight: 500;
    letter-spacing: 1.25px;
  }

  &__control {
    width: 100%;
    padding: rem(12);
    border: 1px solid #e3e3e3;
    background: #fff;

    &--textarea {
      resize: none;
    }
  }

  &__error {
    display: block;
    margin-top: rem(5);
    color: #e74c3c;
    font-size: em(14);
  }
}
