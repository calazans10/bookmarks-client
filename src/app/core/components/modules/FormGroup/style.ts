import styled from 'styled-components';
import { em, rem } from 'polished';

export const Container = styled.div`
  margin-bottom: ${rem('20px')};
`;

export const Label = styled.label`
  display: block;
  margin-bottom: ${rem('5px')};
  font-weight: 500;
  letter-spacing: 1.25px;
`;

export const Input = styled.input`
  width: 100%;
  padding: ${rem('12px')};
  border: ${rem('1px')} solid #e3e3e3;
  background: #fff;
`;

export const ErrorMessage = styled.span`
  display: block;
  margin-top: ${rem('5px')};
  color: #e74c3c;
  font-size: ${em('14px')};
`;
