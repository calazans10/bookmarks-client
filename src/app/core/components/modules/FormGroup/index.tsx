import React, { useState } from 'react';
import { Container, Label, Input, ErrorMessage } from './style';

type FormGroupProps = {
  input: any;
  meta: any;
  label?: string;
  type?: string;
  mask?: string;
  formatChars?: any;
  autoFocus?: boolean;
  disabled?: boolean;
}

const FormGroup = ({
  input,
  meta,
  label = '',
  type = 'tel',
  mask = '',
  formatChars = { 9: '[0-9]', '?': '[0-9]' },
  autoFocus = false,
  disabled = false,
}: FormGroupProps) => {
  const [isFocused, setIsFocused] = useState(false);

  const isInvalid = meta.touched && meta.error && !isFocused;

  const onFocus = () => setIsFocused(true);

  const onBlur = () => setIsFocused(false);

  return (
    <Container>
      <Label htmlFor={input.name}>{label}</Label>
      <Input
        {...input}
        id={input.name}
        type={type}
        mask={mask}
        formatChars={formatChars}
        autoFocus={autoFocus}
        disabled={disabled}
        placeholder={label}
        onFocus={onFocus}
        onBlur={onBlur}
      />
      {isInvalid && <ErrorMessage>{meta.error}</ErrorMessage>}
    </Container>
  );
};

export default FormGroup;
