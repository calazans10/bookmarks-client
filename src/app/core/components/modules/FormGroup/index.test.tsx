import React from 'react';
import { render } from '@testing-library/react';
import FormGroup from '.';

describe('FormGroup', () => {
  const defaultMeta = {
    active: false,
    data: {},
    dirty: false,
    dirtySinceLastSubmit: false,
    error: 'This field is required.',
    invalid: true,
    pristine: true,
    submitFailed: false,
    submitSucceeded: false,
    touched: false,
    valid: false,
    visited: false,
  };

  describe('when is valid', () => {
    it('renders without crashing', () => {
      const props = {
        input: {
          name: 'password',
        },
        meta: { ...defaultMeta, touched: false, type: 'password' },
        label: 'Password',
        type: 'password',
      };

      const { container } = render(<FormGroup {...props} />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('when is invalid', () => {
    it('renders without crashing', () => {
      const props = {
        input: {
          name: 'password',
        },
        meta: { ...defaultMeta, touched: true, type: 'password' },
        label: 'Password',
        type: 'password',
      };

      const { container } = render(<FormGroup {...props} />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
