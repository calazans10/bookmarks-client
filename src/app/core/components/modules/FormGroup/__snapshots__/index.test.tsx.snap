// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormGroup when is invalid renders without crashing 1`] = `
<div
  class="sc-AxjAm gwJtGB"
>
  <label
    class="sc-AxirZ bgWkfA"
    for="password"
  >
    Password
  </label>
  <input
    class="sc-AxiKw bmohav"
    id="password"
    name="password"
    placeholder="Password"
    type="password"
  />
  <span
    class="sc-AxhCb XwNWq"
  >
    This field is required.
  </span>
</div>
`;

exports[`FormGroup when is valid renders without crashing 1`] = `
<div
  class="sc-AxjAm gwJtGB"
>
  <label
    class="sc-AxirZ bgWkfA"
    for="password"
  >
    Password
  </label>
  <input
    class="sc-AxiKw bmohav"
    id="password"
    name="password"
    placeholder="Password"
    type="password"
  />
</div>
`;
